// @volar/vue-typescript/out/plugins/vue-typescript-scripts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const script_1 = require("@volar/vue-code-gen/out/generators/script");
function default_1(lang, scriptRanges, scriptSetupRanges, htmlGen, compilerOptions, cssVars) {
    return {
        getEmbeddedFilesCount(fileName, sfc) {
            return fileName.endsWith('.html') ? 0 : 2;
        },
        getEmbeddedFile(fileName, sfc, i) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            let shimComponentOptionsMode = false;
            if (((_a = compilerOptions.experimentalImplicitWrapComponentOptionsWithDefineComponent) !== null && _a !== void 0 ? _a : 'onlyJs') === 'onlyJs'
                ? lang.value === 'js' || lang.value === 'jsx'
                : !!compilerOptions.experimentalImplicitWrapComponentOptionsWithDefineComponent) {
                shimComponentOptionsMode = 'defineComponent';
            }
            if (((_b = compilerOptions.experimentalImplicitWrapComponentOptionsWithVue2Extend) !== null && _b !== void 0 ? _b : 'onlyJs') === 'onlyJs'
                ? lang.value === 'js' || lang.value === 'jsx'
                : !!compilerOptions.experimentalImplicitWrapComponentOptionsWithVue2Extend) {
                shimComponentOptionsMode = 'Vue.extend';
            }
            // true override 'onlyJs'
            if (compilerOptions.experimentalImplicitWrapComponentOptionsWithDefineComponent === true) {
                shimComponentOptionsMode = 'defineComponent';
            }
            if (compilerOptions.experimentalImplicitWrapComponentOptionsWithVue2Extend === true) {
                shimComponentOptionsMode = 'Vue.extend';
            }
            const codeGen = (0, script_1.generate)(i === 0 ? 'script' : 'template', fileName, (_c = sfc.script) !== null && _c !== void 0 ? _c : undefined, (_d = sfc.scriptSetup) !== null && _d !== void 0 ? _d : undefined, scriptRanges.value, scriptSetupRanges.value, () => htmlGen.value, () => {
                const bindTexts = [];
                for (const cssVar of cssVars.value) {
                    bindTexts.push(cssVar);
                }
                return bindTexts;
            }, shimComponentOptionsMode, ((_e = compilerOptions.experimentalDowngradePropsAndEmitsToSetupReturnOnScriptSetup) !== null && _e !== void 0 ? _e : 'onlyJs') === 'onlyJs'
                ? lang.value === 'js' || lang.value === 'jsx'
                : !!compilerOptions.experimentalDowngradePropsAndEmitsToSetupReturnOnScriptSetup, (_f = compilerOptions.target) !== null && _f !== void 0 ? _f : 3);
            let file;
            if (i === 0) {
                file = {
                    fileName: fileName + '.' + lang.value,
                    content: codeGen.codeGen.getText(),
                    capabilities: {
                        diagnostics: !((_g = sfc.script) === null || _g === void 0 ? void 0 : _g.src),
                        foldingRanges: false,
                        formatting: false,
                        documentSymbol: false,
                        codeActions: !((_h = sfc.script) === null || _h === void 0 ? void 0 : _h.src),
                        inlayHints: !((_j = sfc.script) === null || _j === void 0 ? void 0 : _j.src),
                    },
                    isTsHostFile: true,
                    mappings: codeGen.codeGen.getMappings(),
                    teleportMappings: codeGen.teleports,
                };
            }
            else if (sfc.script || sfc.scriptSetup) {
                file = {
                    fileName: fileName + '.__VLS_script.' + lang.value,
                    content: codeGen.codeGen.getText(),
                    capabilities: {
                        diagnostics: false,
                        foldingRanges: false,
                        formatting: false,
                        documentSymbol: false,
                        codeActions: false,
                        inlayHints: false,
                    },
                    isTsHostFile: true,
                    mappings: codeGen.codeGen.getMappings(),
                    teleportMappings: codeGen.teleports,
                };
            }
            return file;
        },
    };
}
exports.default = default_1;
